pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -0,0 +1,204 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologymanager
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+)
+
+//Manager interface provides methods for Kubelet to manage pod topology hints
+type Manager interface {
+	//Manager implements pod admit handler interface
+	lifecycle.PodAdmitHandler
+	//Adds a hint provider to manager to indicate the hint provider
+	//wants to be consoluted when making topology hints
+	AddHintProvider(HintProvider)
+	//Adds pod to Manager for tracking
+	AddContainer(pod *v1.Pod, containerID string) error
+	//Removes pod from Manager tracking
+	RemoveContainer(containerID string) error
+	//Interface for storing pod topology hints
+	Store
+}
+
+//TopologyHint is a struct containing Socket Mask for a Pod
+type TopologyHint struct {
+	SocketMask socketmask.SocketMask
+}
+
+type manager struct {
+	//The list of components registered with the Manager
+	hintProviders []HintProvider
+	//List of Containers and their Topology Allocations
+	podTopologyHints map[string]containers
+	podMap           map[string]string
+	//Topology Manager Policy
+	policy Policy
+}
+
+//HintProvider interface is to be implemented by Hint Providers
+type HintProvider interface {
+	GetTopologyHints(pod v1.Pod, container v1.Container) ([]TopologyHint, bool)
+}
+
+//Store interface is to allow Hint Providers to retrieve pod affinity
+type Store interface {
+	GetAffinity(podUID string, containerName string) TopologyHint
+}
+
+type containers map[string]TopologyHint
+
+var _ Manager = &manager{}
+
+type policyName string
+
+//NewManager creates a new TopologyManager based on provided policy
+func NewManager(topologyPolicyName string) Manager {
+	klog.Infof("[topologymanager] Creating topology manager with %s policy", topologyPolicyName)
+	var policy Policy
+
+	switch policyName(topologyPolicyName) {
+
+	case PolicyPreferred:
+		policy = NewPreferredPolicy()
+
+	case PolicyStrict:
+		policy = NewStrictPolicy()
+
+	default:
+		klog.Errorf("[topologymanager] Unknow policy %s, using default policy %s", topologyPolicyName, PolicyPreferred)
+		policy = NewPreferredPolicy()
+	}
+
+	var hp []HintProvider
+	pnh := make(map[string]containers)
+	pm := make(map[string]string)
+	manager := &manager{
+		hintProviders:    hp,
+		podTopologyHints: pnh,
+		podMap:           pm,
+		policy:           policy,
+	}
+
+	return manager
+}
+
+func (m *manager) GetAffinity(podUID string, containerName string) TopologyHint {
+	return m.podTopologyHints[podUID][containerName]
+}
+
+func (m *manager) calculateTopologyAffinity(pod v1.Pod, container v1.Container) (TopologyHint, bool) {
+	admitPod := true
+	firstHintProvider := true
+	var containerHints []TopologyHint
+	for _, hp := range m.hintProviders {
+		topologyHints, admit := hp.GetTopologyHints(pod, container)
+		if !admit && topologyHints == nil {
```

Comment:
Agree that the check isn't reflecting what should happen in the event a hint provider doesn't care about a pod. I think the criteria should be admit && topologyHints == nil 

I don't think the return should be nil in the case a hint provider returns false, we should continue to create the hint and store it

Reply:
I think I'd need to see what you are proposing exactly, as I'm still unclear on the flow. It was my mistake to suggest "returns" above as clearly we shouldn't just return from the function but rather continue / break the loop under certain conditions.

Reply:
So 
```if admit && topologyHints == nil {
     //Hint Provider does not care about incoming pod 
     continue
   }
  //calculate TopologyHint
  //if a Hint Provider returns admit = false - admit return will be false
  //if resulting TopologyHint is cross socket admit return will be false
  //else admitPod = true
   
}
}```

Reply:
By "cross socket" you mean having more than 1 socket set in the resulting SocketMask? Why does this result in a pod not being admitted? Doesn't this just mean that the CPU manager is free to choose CPUs from whichever Socket it chooses, and the different hint providers can base their decisions on what do do, knowing full well that this was a possible outcome (since we wouldn't end up with this as a result if all hint providers hadn't indicated that this was an option)?

Reply:
Yes that's what I mean by cross socket. In the strict policy if the pod can only get cross socket allocations, then the pod should not be admitted. 

So it almost results in the same thing - if  hint providers send back a 01 and 10. The hints would be nil and admit would be false. If hint providers send back 01 11 and 10 11, the hint would be 11 and admit would be false. But I think it would be preferable to store the cross socket hint as opposed to nil as it is explicitly telling the hintProviders a hint was generated and this is a result

Reply:
Something about this seems very wrong to me. Why would a hint provider give a hint of 11 as an option if it wasn't ok with the pod being admitted if that is the merged hint resulted in 11. if it wasn't OK with this, then it shouldn't give 11 as an option, forcing the merged hint to become 0 and admit to end up false.

Reply:
As a concrete example, consider a machine with 8 GPUs, 4 connected to each socket, with all GPUs currently available:

* If a container has a request anywhere between 1-4 GPUs, GetTopologyHints() would return a list of TopologyHints with masks {01, 10} set. If 01 is chosen as the stored socketmask by the TopologyManager, then GPUs will be chosen from 1-4 after a call to GetAffinity(), whereas if 10 is chosen as the stored socketmask by the TopologyManager, then GPUs will be chosen from 1-4 after a call to GetAffinity()

* If however, a container has a request for anywhere between 5-8 GPUs, GetTopologyHints would return a single socketmask in TopologyHints as {11} since it doesn't care anymore about a specific socket affinity since GPUs attached to both sockets will always be allocated.

Reply:
These issues have been resolved based on offline discussions about the preferred semantics.

I have created a PR against the branch for this PR to resolve them:
https://github.com/lmdaly/kubernetes/pull/33